core/src/main/java/com/linecorp/armeria/client/HttpClientFactory.java

```diff
@@ -425,8 +431,35 @@ public CompletableFuture<?> closeAsync() {
     }
 
     private void closeAsync(CompletableFuture<?> future) {
+        if (drainFuture != null) {
+            final boolean cancelled;
+            if (drainFuture.getDelay(TimeUnit.MICROSECONDS) > drainDurationMillis) {
+                // Maybe reschedule below.
+                cancelled = drainFuture.cancel(false);
+            } else {
+                cancelled = false;
+            }
+
+            if (cancelled) {
+                drainFuture = null;
+            } else {
+                // Drain is already scheduled to finish earlier.
+                return;
+            }
+        }
+
+        if (drainDurationMillis > 0) {
+            drainFuture = workerGroup.schedule(() -> finish(future),
+                                               drainDurationMillis, TimeUnit.MICROSECONDS);
+        } else {
+            finish(future);
+        }
+    }
+
+    private void finish(CompletableFuture<?> future) {
         final List<CompletableFuture<?>> dependencies = new ArrayList<>(pools.size());
         for (final Iterator<HttpChannelPool> i = pools.values().iterator(); i.hasNext();) {
+            // Close all channels gracefully just like Http2serverConnectionHandler
```

Comment:
This doesn't close the channels gracefully.
Instead of modifying this factory, we should implement the feature in the `Http2ClientConnectionHandler`.
Let's add `gracefulConnectionShutdownHandler` to the `Http2ClientConnectionHandler`. We can refer to [Http2ServerConnectionHandler](https://github.com/line/armeria/blob/6e485eb4c578271818e5631335d33f0ed96e2859/core/src/main/java/com/linecorp/armeria/server/Http2ServerConnectionHandler.java#L37).

When the `Http2ClientConnectionHandler.close()` is called, we can defer the closing task to the `gracefulConnectionShutdownHandler` instead of closing the channel right away.

Reply:
Thank you for the detailed explanation 👍

Reply:
Btw, this PR seems to have missed many points. I'll make sure to do more checks before submitting a PR 😄